### ISSUE TITLE
Add option to cache data at nodes of learning networks to avoid repeating operations (transform, predict, etc)

### ISSUE THREAD

**Issue by Leonardbcm** [alan-turing-institute/MLJ.jl#702]
Hi everyone,

I am manipulating pipelines of unsupervised transformers. I want to stack transformers linearly and use them all at once using the pipeline object. In my situation, the tranformation time of each individual transformer is much higher than the fitting time.

I have noticed that while fitting each transformer, the pipeline re-computes from scratch all the previous transformations.
Here is a quick example:
```julia
using MLJ

X = rand(Float64, 100, 4)

#### Define a first transformer
mutable struct Transformer <: Unsupervised
    a::Int
end

Transformer(; a::Int=2) = Transformer1(a) 

function MLJ.fit(model::Transformer, verbosity::Int, X)
    @info "--> Fitting transformer $(model.a)"
    fitresult = model.a
    cache = nothing
    report = nothing
    return fitresult, cache, report
end

function MLJ.transform(model::Transformer, fitresult, X)
    @info "--> Transforming transformer $(model.a)"
    return fitresult * X
end

t1 = Transformer()
mt1 = machine(t1, X)

fit!(mt1; verbosity=4)
transform(mt1, X)

#### Create a pipeline
pipe = @pipeline Transformer(1) Transformer(2) Transformer(3) Transformer(4)
mach_pipe = machine(pipe, X)

fit!(mach_pipe; verbosity=4)
transform(mach_pipe, X);
``` 

> julia> fit!(mach_pipe; verbosity=4)
[ Info: Training Machine{Pipeline288} @262.
[ Info: Training Machine{Transformer} @165.
[ Info: --> Fitting transformer 1
[ Info: --> Transforming transformer 1
[ Info: Training Machine{Transformer} @449.
[ Info: --> Fitting transformer 2
[ Info: --> Transforming transformer 1
[ Info: --> Transforming transformer 2
[ Info: Training Machine{Transformer} @332.
[ Info: --> Fitting transformer 3
[ Info: --> Transforming transformer 1
[ Info: --> Transforming transformer 2
[ Info: --> Transforming transformer 3
[ Info: Training Machine{Transformer} @272.
[ Info: --> Fitting transformer 4
Machine{Pipeline288} @262 trained 1 time.
  args:
    1:  Source @285 ⏎ `AbstractArray{Continuous,2}`

We see that after fitting `transformer 2`, the pipeline re-calls `transform` from the first transformer. It is supposed to already have in store the transformation results from the `transformer 1` because this data has been used to `fit` `transformer 2`. 
The same happens for `transformer 3` as `transform` is called again for `transformer 1` and `transformer 2`.

Is there a particular reason for this behavior? How can I avoid re-performing the transformation step for every new transformers? 

Thanks for your answers!

**Comment by ablaom** [comment]
Transformations are indeed called each time because pipelines and other learning networks do not cache data at each node. All operations in a learning network (transform, predict, and so forth) are lazy. And I think for most use-cases this is the preferred default behaviour. Making copies of the data at each node would be inappropriate for very large data sets, and, generally, training is is the bottleneck - not transforming/predicting. 

Caching data at  nodes could be made an option, and I've changed the issue title to reflect your request.

**Comment by Leonardbcm** [comment]
Okay thanks!

**Comment by ablaom** [comment]
Actually, caching data at a machine (whether stand-alone or in a learning network) is implemented and the default behaviour in MLJBase 0.17.0. See [this PR](https://github.com/alan-turing-institute/MLJBase.jl/pull/501).

To turn caching off, see here: https://github.com/alan-turing-institute/MLJ.jl/blob/mlj2/docs/src/machines.md#lowering-memory-demands

The real benefit comes when models implement an optional data front-end (to prevent things like repeated table -> matrix conversions). See here:

https://github.com/alan-turing-institute/MLJ.jl/blob/mlj2/docs/src/adding_models_for_general_use.md#implementing-a-data-front-end

**Comment by ablaom** [comment]
Will be live in MLJ 0.16.
